kyo-pure/shared/src/main/scala/kyo/core.scala

```diff
@@ -0,0 +1,214 @@
+            v match
+                case kyo: Suspend[MX, Any, T, S] @unchecked =>
+                    new Continue[MX, Any, T, S & S2](kyo):
+                        def apply(v: Any, s: Safepoint[S & S2], l: Locals.State) =
+                            evalLoop(s, f(s, () => kyo(v, s.asInstanceOf[Safepoint[S]], l)))
+                        end apply
+                case v =>
+                    f(s, () => v.asInstanceOf[T])
+        evalLoop(Safepoint.noop, v)
+    end eval
+
+    inline def transform[T, S, U, S2](v: T < S)(inline k: T => (U < S2)): U < (S & S2) =
+        def transformLoop(v: T < S): U < (S & S2) =
+            v match
+                case kyo: Suspend[MX, Any, T, S] @unchecked =>
+                    new Continue[MX, Any, U, S & S2](kyo):
+                        def apply(v: Any, s: Safepoint[S & S2], l: Locals.State) =
+                            val r = kyo(v, s.asInstanceOf[Safepoint[S]], l)
+                            if s.preempt() then
+                                s.suspend(transformLoop(r))
+                            else
+                                transformLoop(r)
+                            end if
+                        end apply
+                case v =>
+                    k(v.asInstanceOf[T])
+        transformLoop(v)
+    end transform
+
+    extension [E <: Effect[E]](e: E)
+        inline def handle[State, Result[_], T, S, S2](
+            handler: ResultHandler[State, e.Command, E, Result, S]
+        )(
+            state: State,
+            value: T < (E & S2)
+        )(using inline tag: Tag[E], inline flat: Flat[T < (E & S)]): Result[T] < (S & S2) =
+            def _handleLoop(st: State, value: T < (E & S & S2)): Result[T] < (S & S2) =
+                handleLoop(st, value)
+            @tailrec def handleLoop(st: State, value: T < (E & S & S2)): Result[T] < (S & S2) =
+                value match
+                    case kyo: Suspend[e.Command, Any, T, S2] @unchecked
+                        if tag == kyo.tag && handler.accepts(st, kyo.command) =>
+                        handler.resume(st, kyo.command, kyo) match
+                            case r: handler.Resume[T, S & S2] @unchecked =>
+                                handleLoop(r.st, r.v)
+                            case kyo: Kyo[Result[T] | handler.Resume[T, S2], S & S2] @unchecked =>
+                                resultLoop(kyo)
+                            case r =>
+                                r.asInstanceOf[Result[T]]
+                        end match
+                    case kyo: Suspend[MX, Any, T, E & S & S2] @unchecked =>
+                        new Continue[MX, Any, Result[T], S & S2](kyo):
+                            def apply(v: Any, s: Safepoint[S & S2], l: Locals.State) =
+                                _handleLoop(st, kyo(v, s, l))
+                    case v =>
+                        handler.done(st, v.asInstanceOf[T])
```

Comment:
I made a few minor optimizations to reduce bytecode size. Instead of matching the pure type at the end, I changed the code to use a type cast since it's the last type check.